### DISCUSSION TITLE
[Feature] Built-in primary key auto-generation strategy

### DISCUSSION THREAD

**Discussion by coldsmirk** [SeaQL/sea-orm#1992] [Ideas]
## Motivation

If I don't use the auto-increment method of the database's primary key, then I have to generate the primary key automatically when inserting data in the program. However, this is a highly repetitive and irrelevant task. It would be ideal if the framework could handle it instead.

## Proposed Solutions

We can set a primary key generation strategy for orm framwork, or the framework itself provides several ready-to-use strategies for users to choose directly.

## Current Workarounds

Manually generate a primary key, then set it to ActiveModel and save it to the database.

**Comment by tyt2y3** [comment]
That's a very interesting idea, what are some ready-to-use strategies you have in mind?

**Comment by coldsmirk** [comment]
Thanks.

First of all, from my previous experience using Java ORM frameworks, UUID would be a kind (built-in in Hibernate) that may not be commonly used but can still be used in simple scenarios, although it is not recommended to use UUID as the primary key for tables.

Secondly, another generation strategy is similar to Facebook's Snowflake ID (built-in in MybatisPlus). I saw a good library for this in Rust: [`idgenerator`](https://crates.io/crates/idgenerator).

**Comment by tyt2y3** [comment]
Thank you for your suggestions. They all seemed very interesting. Without `associated_type_defaults`, customization of the default behaviour inside `EntityTrait` would require extra code in the Entity file. Having an extra `EntityBehaviour` trait seems to be too much.

Along with #1956 the problem is whether we add an default type parameter to entity:

```rust
trait EntityTrait<B: EntityBehaviour = DefaultEntityBehaviour>
```

Or attach it to `ActiveModelBehaviour`

```rust
trait EntityTrait {
    ...
    type A: ActiveModelBehaviour;
}
```

The latter would require a breaking change, but I feel like it's a cleaner solution for the long term.

**Comment by coldsmirk** [comment]
I also think it would be better to attach it to `ActiveModelBehaviour`. We can use `ActiveModelBehavior` to implement default methods to provide a default primary key generation strategy. If the user wants to customize the primary key generation strategy, they can override the methods implemented by the `ActiveModelBehavior`.
